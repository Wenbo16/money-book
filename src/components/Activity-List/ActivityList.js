import React from "react";
import { IonIcon, addIcons } from "react-svg-ionicons";
import bundle from "react-svg-ionicons/bundles/all";

import PropTypes from "prop-types";
import { Colors } from "../../utility";

addIcons(bundle);

const ActivityList = ({ items, onModifyItem, onDeleteItem }) => {
  return (
    <ul className="list-group list-group-flush">
      <li
        className="list-group-item d-flex justify-content-between align-items-center"
        style={{ borderWidth: "0 0 2px" }}
      >
        <span className="col-1 font-weight-bold">类型</span>
        <span className="col-5 font-weight-bold">类容</span>
        <span className="col-2 font-weight-bold">价格</span>
        <span className="col-2 font-weight-bold">时间</span>
        <span className="col-1 font-weight-bold">修改</span>
        <span className="col-1 font-weight-bold">删除</span>
      </li>
      {items.map((item) => (
        <li
          className="list-group-item d-flex 
            justify-content-between align-items-center"
          key={item.id}
        >
          <span className="col-1">
            <IonIcon
              className="rounded-circle"
              fontSize="30px"
              style={{ backgroundColor: Colors.blue, padding: "5px" }}
              color={"#fff"}
              name={item.category.iconName}
            />
          </span>
          <span className="col-5">{item.title}</span>
          <span className="col-2 font-weight-bold">
            {item.category.type === "income" ? "+" : "-"}
            {item.amount}元
          </span>
          <span className="col-2">{item.date}</span>
          <a
            className="col-1"
            role="button"
            onClick={(event) => {
              onModifyItem(item);
            }}
          >
            <IonIcon
              className="rounded-circle"
              fontSize="30px"
              style={{ backgroundColor: Colors.green, padding: "5px" }}
              color={"#fff"}
              name="create"
            />
          </a>
          <a
            className="col-1"
            role="button"
            onClick={(event) => {
              onDeleteItem(item);
            }}
          >
            <IonIcon
              className="rounded-circle"
              fontSize="30px"
              style={{ backgroundColor: Colors.red, padding: "5px" }}
              color={"#fff"}
              name="close"
            />
          </a>
        </li>
      ))}
    </ul>
  );
};

ActivityList.propTypes = {
  items: PropTypes.array.isRequired,
  onModifyItem: PropTypes.func.isRequired,
  onDeleteItem: PropTypes.func.isRequired,
};

export default ActivityList;
