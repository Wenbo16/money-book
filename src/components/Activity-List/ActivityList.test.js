import React from 'react';
import { shallow } from 'enzyme';
import Ionicon from 'react-ionicons';
import ActivityList from './ActivityList';
import { testItems, testCategories } from '../../testData';
import { flattenArr } from '../../utility';

/**
 * 1. 传入特定数组，是否渲染对应长度的条目
 * 2. 每个条目是否渲染特定组件和内容
 * 3. 点击按钮是否触发特定回调
 */

const itemsWithCategory = testItems.map((item) => {
  item.category = flattenArr(testCategories)[item.cid];
  return item;
});

const props = {
  items: itemsWithCategory,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn(),
};
let wrapper;
describe('test ActivityList component', () => {
  // excute before each it
  beforeEach(() => {
    wrapper = shallow(<ActivityList {...props} />);
  });

  it('should render the component to match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct amount items length', () => {
    expect(wrapper.find('.list-group-item').length).toEqual(
      itemsWithCategory.length
    );
  });

  it('should render correct icon and amount for each item', () => {
    const iconList = wrapper.find('.list-group-item').first().find(Ionicon);
    expect(iconList.length).toEqual(3);
    expect(iconList.first().props().icon).toEqual(
      itemsWithCategory[0].category.iconName
    );
  });

  it('should trigger the correct function callbacks', () => {
    const firstItem = wrapper.find('.list-group-item').first();
    firstItem.find('a').first().simulate('click');
    expect(props.onModifyItem).toHaveBeenCalledWith(itemsWithCategory[0]);
    firstItem.find('a').last().simulate('click');
    expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategory[0]);
  });
});
