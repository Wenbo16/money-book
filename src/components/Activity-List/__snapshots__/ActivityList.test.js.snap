// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ActivityList component should render the component to match snapshot 1`] = `undefined`;
